optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/termination/TerminationFactory.java

```diff
@@ -0,0 +1,160 @@
+            terminationList.add(new UnimprovedStepCountTermination(terminationConfig.getUnimprovedStepCountLimit()));
+        }
+
+        terminationList.addAll(buildInnerTermination(configPolicy));
+        return buildTerminationFromList(terminationList);
+    }
+
+    protected List<Termination> buildTimeBasedTermination(HeuristicConfigPolicy configPolicy) {
+        List<Termination> terminationList = new ArrayList<>();
+        Long timeMillisSpentLimit = terminationConfig.calculateTimeMillisSpentLimit();
+        if (timeMillisSpentLimit != null) {
+            terminationList.add(new TimeMillisSpentTermination(timeMillisSpentLimit));
+        }
+        Long unimprovedTimeMillisSpentLimit = terminationConfig.calculateUnimprovedTimeMillisSpentLimit();
+        if (unimprovedTimeMillisSpentLimit != null) {
+            if (terminationConfig.getUnimprovedScoreDifferenceThreshold() == null) {
+                terminationList.add(new UnimprovedTimeMillisSpentTermination(unimprovedTimeMillisSpentLimit));
+            } else {
+                ScoreDefinition scoreDefinition = configPolicy.getScoreDefinition();
+                Score unimprovedScoreDifferenceThreshold_ =
+                        scoreDefinition.parseScore(terminationConfig.getUnimprovedScoreDifferenceThreshold());
+                if (unimprovedScoreDifferenceThreshold_.compareTo(scoreDefinition.getZeroScore()) <= 0) {
+                    throw new IllegalStateException("The unimprovedScoreDifferenceThreshold ("
+                            + terminationConfig.getUnimprovedScoreDifferenceThreshold() + ") must be positive.");
+
+                }
+                terminationList.add(new UnimprovedTimeMillisSpentScoreDifferenceThresholdTermination(
+                        unimprovedTimeMillisSpentLimit, unimprovedScoreDifferenceThreshold_));
+            }
+        } else if (terminationConfig.getUnimprovedScoreDifferenceThreshold() != null) {
+            throw new IllegalStateException("The unimprovedScoreDifferenceThreshold ("
+                    + terminationConfig.getUnimprovedScoreDifferenceThreshold()
+                    + ") can only be used if an unimproved*SpentLimit ("
+                    + unimprovedTimeMillisSpentLimit + ") is used too.");
+        }
+
+        return terminationList;
+    }
+
+    protected List<Termination> buildInnerTermination(HeuristicConfigPolicy configPolicy) {
+        if (ConfigUtils.isEmptyCollection(terminationConfig.getTerminationConfigList())) {
+            return Collections.emptyList();
+        }
+
+        return terminationConfig.getTerminationConfigList().stream()
+                .map(config -> TerminationFactory.create(config).buildTermination(configPolicy))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    protected Termination buildTerminationFromList(List<Termination> terminationList) {
+        if (terminationList.size() == 1) {
```

Comment:
I'd change the logic here. Start with `isEmpty()`, then check for `size == 1`.
That way, you avoid calling `size()` twice, and also messing with boundary conditions (`>`).